### ISSUE TITLE
Contacts.getAll() crashes Android app when permissions are not granted

### ISSUE THREAD

**Issue by aelesia** [morenoh149/react-native-contacts#516]
# Expected Behaviour:

If permissions are not granted, the callback should be called, with the `error` field being non-null/undefined.

```
    Contacts.getAll((err, contacts) => {
        // err should be defined
    })
```

# Actual Behaviour:

App crashes completely. React Native does not even have the chance to catch the error.

# My code:

```typescript
async function requestPermissions() {
  let status = await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS)
  console.log('status', status)
}

async function contactsGetAll(): Promise<Contact[]> {
  await requestPermissions()
  console.log('permissions')
  return new Promise((res, rej) => {
    Contacts.getAll((err, contacts) => {
      console.log('err', err)
      if (err) {  rej(err) } 
     else { res(contacts) }
    })
  })
}

contacts: Contact[] = await contactsGetAll()
```

# console.log output:

```typescript
status: never_ask_again
permissions
// crashes
```

# Workaround fix:

Check permissions before calling `Contacts.getAll()`

```typescript
  let status = await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS)
  if (status === 'denied' || status === 'never_ask_again') {
    throw Error('Permissions not granted to access Contacts')
  }
Contacts.getAll((err, contacts) => {})
```

**Comment by morenoh149** [comment]
I'm not sure if there's anything we can/should do about this. Checking for permissions before performing privileged actions seems fine to me.

**Comment by xapai** [comment]
I think, just crashing silently is not a way to go either. Maybe provide some kind of error handling which will prevent crash. I'm not the most Java literate person at a moment, but will try to provide some PR with a fix, if be able to scrape some free time.

**Comment by aelesia** [comment]
> I'm not sure if there's anything we can/should do about this. Checking for permissions before performing privileged actions seems fine to me.

There is inconsistent behaviour on Android vs iOS though.

In iOS, permissions aren't granted, it will be handled in the error block:

```
    Contacts.getAll((err, contacts) => {
        // iOS permission errors will be caught here
    })
```

In Android, if permissions are not granted, errors do not get caught by this block. It crashes the entire app entirely.

__________

1) Why provide an error handler for when the error handler only catches iOS errors and not Android errors? It's misleading to developers to see an error handler block that does not catch errors. Wouldn't it be optimal for both Android and iOS errors to be caught in the same block?

2) It does not seem appropriate that calling a ReactNative function can cause an entire Android app to crash. I personally feel that errors should be handled at the ReactNative boundary.

**Comment by anandhakumarpalanisamy** [comment]
**I solved it by the using the below piece of "try catch" code logic:** 

The app will not crash, since all the  cases are handled .

```typescript

try {
        const andoidContactPermission = await PermissionsAndroid.request(
          PermissionsAndroid.PERMISSIONS.READ_CONTACTS,
          {
            title: "Contacts Permission",
            message:
              "This app would like to view your contacts.",
            buttonNeutral: "Ask Me Later",
            buttonNegative: "Cancel",
            buttonPositive: "OK"
          }
        );
        if (andoidContactPermission === PermissionsAndroid.RESULTS.GRANTED) {
          console.log("Contacts Permission granted");
          Contacts.getAll((andoidContactPermission, contacts) => {
              console.log(contacts);
          });
        } else {
          console.log("Contacts permission denied");
        }
      } catch (err) {
        console.log(err);
 }
    
    

```

Hope this helps :+1:

**Comment by ZainaliSyed** [comment]
It's still crashing working fine on some phones but on few phones it's crashing 

Crashing on following devices 
Samsung S8 plus 
Android Version 5 : Hawaii P8 lite

**Comment by abbasmoosavi** [comment]
in android manifest file use
```java
<uses-permission android:name="android.permission.READ_CONTACTS" />
    <uses-permission
        android:name="android.permission.READ_PROFILE"
        android:maxSdkVersion="22"
        tools:targetApi="donut" />
```

**Comment by aelesia** [comment]
Btw, just sharing my code if anyone just wants a quick copy and paste:

Usage:
```typescript
Contacts.getAll().then(contacts => {
  contacts.forEach(contact => {
    console.log(contact)
    console.log(contact.givenName)
    console.log(contact.phoneNumbers)
  })
})
```

Code:

**Contacts.ts**
```typescript
// Wrapper for permissions as original function does not throw
const requestPermissions = async (): Promise<void> => {
  // If app would like to display the rational for requesting permissions,
  // then place it as second param in PermissionsAndroid.request().
  if (Platform.OS === 'android') {
    let status = await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS)
    if (status === 'denied' || status === 'never_ask_again') {
      throw Error('Permission not granted to access contacts')
    }
  }
}

// Wrapper for requesting contacts as original function returns callback instead of Promise
const getAll = async (): Promise<Contact[]> => {
  await requestPermissions()
  return new Promise((res, rej) => {
    RNContacts.getAll((err, contacts) => {
      if (err) {
        rej(new PermissionContactsErr())
      } else {
        res(contacts)
      }
    })
  })
}

export const Contacts = { getAll }
```

**Comment by cvharris** [comment]
I ran into this too and it was super weird, I found you need both `WRITE_CONTACTS` and `READ_CONTACTS` regardless of API version. The README sections them off as if READ was a part of WRITE for Api 23+, that's what confused me at least.

**Comment by morenoh149** [comment]
@cvharris not sure. If you want to update the readme that would be welcome.

**Comment by chawkibenalaya** [comment]
My PermissionsAndroid is granted and i cannot catch the error , still have the crash with API 22 . Any Suggestions Guys ?

**Comment by abbasmoosavi** [comment]
> My PermissionsAndroid is granted and i cannot catch the error , still have the crash with API 22 when i make a `getAll` call . Any Suggestions Guys ?

You must use read profile permission in android platform.

**Comment by eldirb21** [comment]
<uses-permission tools:node="remove" android:name="android.permission.SYSTEM_ALERT_WINDOW" />
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.READ_PROFILE" />
    <uses-permission android:name="android.permission.WRITE_CONTACTS" />
    <uses-permission android:name="android.permission.READ_CONTACTS" />

 i do not understand why i need READ_PROFILE ?

**Comment by Michaelvons** [comment]
> **I solved it by the using the below piece of "try catch" code logic:**
> 
> The app will not crash, since all the cases are handled .
> 
> ```ts
> try {
>         const andoidContactPermission = await PermissionsAndroid.request(
>           PermissionsAndroid.PERMISSIONS.READ_CONTACTS,
>           {
>             title: "Contacts Permission",
>             message:
>               "This app would like to view your contacts.",
>             buttonNeutral: "Ask Me Later",
>             buttonNegative: "Cancel",
>             buttonPositive: "OK"
>           }
>         );
>         if (andoidContactPermission === PermissionsAndroid.RESULTS.GRANTED) {
>           console.log("Contacts Permission granted");
>           Contacts.getAll((andoidContactPermission, contacts) => {
>               console.log(contacts);
>           });
>         } else {
>           console.log("Contacts permission denied");
>         }
>       } catch (err) {
>         console.log(err);
>  }
>     
>     
> ```
> 
> Hope this helps 👍

The newer way of calling Contacts.getAll() is now as follows:


```
        Contacts.getAll()
          .then(contacts => {
            // work with contacts
            console.log(contacts);
          })
          .catch(e => {
            console.log(e);
          });

```

**Comment by mohammadgharouni** [comment]
i'm on version 7.0.7 and adding `<uses-permission android:name="android.permission.READ_PROFILE" />` fixed my issue on react-native 0.70.6
